google/cloud/internal/oauth2_google_credentials.cc

```diff
@@ -0,0 +1,291 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/oauth2_google_credentials.h"
+#include "google/cloud/internal/compute_engine_util.h"
+#include "google/cloud/internal/filesystem.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/make_jwt_assertion.h"
+#include "google/cloud/internal/oauth2_anonymous_credentials.h"
+#include "google/cloud/internal/oauth2_authorized_user_credentials.h"
+#include "google/cloud/internal/oauth2_compute_engine_credentials.h"
+#include "google/cloud/internal/oauth2_credentials.h"
+#include "google/cloud/internal/oauth2_google_application_default_credentials_file.h"
+#include "google/cloud/internal/oauth2_service_account_credentials.h"
+#include "google/cloud/internal/throw_delegate.h"
+#include "absl/memory/memory.h"
+#include <nlohmann/json.hpp>
+#include <fstream>
+#include <iterator>
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+char constexpr kAdcLink[] =
+    "https://developers.google.com/identity/protocols/"
+    "application-default-credentials";
+
+// Parses the JSON at `path` and creates the appropriate
+// Credentials type.
+//
+// If `service_account_scopes` or `service_account_subject` are specified, the
+// file at `path` must be a JSON service account. If a different type of
+// credential file is found, this function returns nullptr to indicate a service
+// account file wasn't found.
+StatusOr<std::unique_ptr<Credentials>> LoadCredsFromPath(
+    std::string const& path, bool non_service_account_ok,
+    absl::optional<std::set<std::string>> service_account_scopes,
+    absl::optional<std::string> service_account_subject,
+    Options const& options) {
+  std::ifstream ifs(path);
+  if (!ifs.is_open()) {
+    // We use kUnknown here because we don't know if the file does not exist, or
+    // if we were unable to open it for some other reason.
+    return Status(StatusCode::kUnknown, "Cannot open credentials file " + path);
+  }
+  std::string contents(std::istreambuf_iterator<char>{ifs}, {});
+  auto cred_json = nlohmann::json::parse(contents, nullptr, false);
+  if (cred_json.is_discarded()) {
```

Comment:
Not a full review, just a comment. I think we mostly stopped using `.is_discarded()` in #8033.

There are still some uses of `.is_discarded()` in `storage`, though. I will defer to @coryan for when to use one vs. the other.
https://github.com/googleapis/google-cloud-cpp/search?q=is_discarded%28%29